### PR TITLE
[CBRD-25080] Add cost formula for LIMIT clause.

### PR DESCRIPTION
**Pull request by shparkcubrid**
http://jira.cubrid.org/browse/CBRD-25080

- Set the cardinality of the subquery having LIMIT clause to value of LIMIT clause.
- If there is a LIMIT clause, it optimizes more rule-based.
 - Rule-based optimization gives priority to execution plans that retrieve effectively the first few rows, such as order by skip and sort-limit.

### CODE REVIEW

File: src/optimizer/query_graph.c [hunk 1]
```diff
@@ -1305,6 +1306,14 @@ qo_add_node (PT_NODE * entity, QO_ENV * env)
 	      if (xasl)
 		{
 		  QO_NODE_NCARD (node) = (unsigned long) xasl->cardinality;
+
+		  /* check limit value */
+		  limit_val = pt_get_query_limit (env->parser, entity->info.spec.derived_table);
+		  if (limit_val > 0 && (unsigned long) limit_val < QO_NODE_NCARD (node))
+		    {
+		      QO_NODE_NCARD (node) = (unsigned long) limit_val;
```
Comment (youngjinj) [inline src/optimizer/query_graph.c:1314]: When limit is used with offset, card becomes the sum of row_count and offset.
```
select /*+ recompile */ count(*) 
from (select /*+ no_merge */ * from t111 where col1 is not null limit 20000 offset 10000) a, 
        (select /*+ no_merge */ * from t111 where col3 = 1) b,
        t111 c 
where c.col3 = 1 
   and a.col1 = b.col1 
   and a.col1 = c.col1
   and b.col1 = c.col1;
```
```
Query plan:

nl-join (inner join)
    edge:  term[0] AND term[1]
    outer: idx-join (inner join)
               outer: sscan
                          class: b node[1]
                          cost:  76 card 10000
               inner: iscan
                          class: c node[2]
                          index: idx1 term[2] AND term[3] (covers)
                          cost:  3 card 10000
               cost:  5103 card 1000
    inner: sscan
               class: a node[0]
               sargs: term[0] AND term[1]
               cost:  228 card 30000 <---------- expected resut: 20000
    cost:  248403 card 1
```
Please refer to tastecases for CBRD-25080.txt.

Comment (shparkcubrid) [reply src/optimizer/query_graph.c:1314]: I fixed it.

**Review by xmilex-git — APPROVED** [review]

**Review by youngjinj — APPROVED** [review]

File: src/optimizer/query_planner.c [hunk 2]
```diff
@@ -71,10 +71,11 @@
 #define QO_CPU_WEIGHT   0.0025
 #define MJ_CPU_OVERHEAD_FACTOR   20
 #define ISCAN_IO_HIT_RATIO   0.5
-#define SSCAN_DEFULT_CARD 1000
+#define SSCAN_DEFULT_CARD 100
```
Comment (beyondykk9) [inline src/optimizer/query_planner.c]: you misspelled SSCAN_DEFAULT_CARD ?

File: src/parser/parser_support.c [hunk 3]
```diff
@@ -153,7 +153,8 @@ static PT_NODE *pt_make_query_user_groups (PARSER_CONTEXT * parser, const char *
 static void pt_help_show_create_table (PARSER_CONTEXT * parser, PT_NODE * table_name, string_buffer & strbuf);
 static int pt_get_query_limit_from_orderby_for (PARSER_CONTEXT * parser, PT_NODE * orderby_for, DB_VALUE * upper_limit,
 						bool * has_limit);
-static int pt_get_query_limit_from_limit (PARSER_CONTEXT * parser, PT_NODE * limit, DB_VALUE * limit_val);
+static int pt_get_query_limit_from_limit (PARSER_CONTEXT * parser, PT_NODE * limit, DB_VALUE * limit_val,
```
Comment (beyondykk9) [inline src/parser/parser_support.c:156]: I suggest to rename to **pt_get_query_limit_from_limit_clause** for readability.

File: src/parser/parser_support.c [hunk 4]
```diff
@@ -9625,6 +9633,35 @@ pt_get_query_limit_value (PARSER_CONTEXT * parser, PT_NODE * query, DB_VALUE * l
...
 }
 
+/*
+ * pt_get_query_limit () - get the limit value from a query
+ * return : error code or NO_ERROR
+ * parser (in)	      : parser context
+ * query (in)	      : query
+ */
+DB_BIGINT
+pt_get_query_limit (PARSER_CONTEXT * parser, PT_NODE * query)
```
Comment (beyondykk9) [inline src/parser/parser_support.c]: How about renaming to **pt_get_query_limit_from_query** for readability?

File: src/parser/parser_support.c [hunk 5]
```diff
@@ -9503,7 +9504,7 @@ pt_make_query_show_collation (PARSER_CONTEXT * parser, int like_where_syntax, PT
  *  specified as LIMIT :offset, :row_count this function returns :offset + :row_count)
  */
 static int
-pt_get_query_limit_from_limit (PARSER_CONTEXT * parser, PT_NODE * limit, DB_VALUE * limit_val)
+pt_get_query_limit_from_limit (PARSER_CONTEXT * parser, PT_NODE * limit, DB_VALUE * limit_val, bool add_offset)
```
Comment (beyondykk9) [inline src/parser/parser_support.c:9507]: How about renaming to **pt_get_query_limit_from_limit_clause** for readability?

File: src/optimizer/query_planner.c (hunk 2 above)
Comment (shparkcubrid) [inline src/optimizer/query_planner.c]:
```suggestion
#define SSCAN_DEFAULT_CARD 100
```

File: src/optimizer/query_planner.c [hunk 6]
```diff
@@ -2988,7 +2989,7 @@ qo_nljoin_cost (QO_PLAN * planp)
     {
       /* if inner is seq scan, it is calculated by default card. */
       /* This prevents the worst plan if the cardinality is calculated to be less than the actual value. */
-      inner_io_cost = MAX (guessed_result_cardinality, SSCAN_DEFULT_CARD) * inner->variable_io_cost;
+      inner_io_cost = (guessed_result_cardinality + SSCAN_DEFULT_CARD) * inner->variable_io_cost;
```
Comment (shparkcubrid) [inline src/optimizer/query_planner.c]:
```suggestion
      inner_io_cost = (guessed_result_cardinality + SSCAN_DEFAULT_CARD) * inner->variable_io_cost;
```

File: src/optimizer/query_planner.c (hunk 2 above)
Comment (shparkcubrid) [reply src/optimizer/query_planner.c]: I fixed it.

File: src/optimizer/query_graph.c [hunk 7]
```diff
@@ -1305,6 +1306,14 @@ qo_add_node (PT_NODE * entity, QO_ENV * env)
 	      if (xasl)
 		{
 		  QO_NODE_NCARD (node) = (unsigned long) xasl->cardinality;
+
+		  /* check limit value */
+		  limit_val = pt_get_query_limit (env->parser, entity->info.spec.derived_table);
```
Comment (shparkcubrid) [inline src/optimizer/query_graph.c]:
```suggestion
		  limit_val = pt_get_query_limit_for_query (env->parser, entity->info.spec.derived_table);
```

File: src/parser/parser.h [hunk 8]
```diff
@@ -651,7 +651,8 @@ extern "C"
 
   extern PT_NODE *pt_find_node_type_pre (PARSER_CONTEXT * parser, PT_NODE * node, void *arg, int *continue_walk);
   extern PT_NODE *pt_find_op_type_pre (PARSER_CONTEXT * parser, PT_NODE * node, void *arg, int *continue_walk);
-  extern int pt_get_query_limit_value (PARSER_CONTEXT * parser, PT_NODE * query, DB_VALUE * limit_val);
+  extern int pt_get_query_limit_value (PARSER_CONTEXT * parser, PT_NODE * query, DB_VALUE * limit_val, bool add_offset);
+  extern DB_BIGINT pt_get_query_limit (PARSER_CONTEXT * parser, PT_NODE * query);
```
Comment (shparkcubrid) [inline src/parser/parser.h]:
```suggestion
  extern DB_BIGINT pt_get_query_limit_for_query (PARSER_CONTEXT * parser, PT_NODE * query);
```

File: src/parser/parser_support.c [hunk 9]
```diff
@@ -9625,6 +9633,35 @@ pt_get_query_limit_value (PARSER_CONTEXT * parser, PT_NODE * query, DB_VALUE * l
   return NO_ERROR;
 }
 
+/*
+ * pt_get_query_limit () - get the limit value from a query
```
Comment (shparkcubrid) [inline src/parser/parser_support.c]:
```suggestion
 * pt_get_query_limit_from_query () - get the limit value from a query
```

File: src/parser/parser_support.c (hunk 4 above)
Comment (shparkcubrid) [inline src/parser/parser_support.c]:
```suggestion
pt_get_query_limit_from_query (PARSER_CONTEXT * parser, PT_NODE * query)
```

File: src/optimizer/query_graph.c [hunk 10]
```diff
@@ -1305,6 +1306,14 @@ qo_add_node (PT_NODE * entity, QO_ENV * env)
 	      if (xasl)
 		{
 		  QO_NODE_NCARD (node) = (unsigned long) xasl->cardinality;
+
+		  /* check limit value */
+		  limit_val = pt_get_query_limit_for_query (env->parser, entity->info.spec.derived_table);
```
Comment (shparkcubrid) [inline src/optimizer/query_graph.c]:
```suggestion
		  limit_val = pt_get_query_limit_from_query (env->parser, entity->info.spec.derived_table);
```

File: src/parser/parser.h [hunk 11]
```diff
@@ -651,7 +651,8 @@ extern "C"
 
   extern PT_NODE *pt_find_node_type_pre (PARSER_CONTEXT * parser, PT_NODE * node, void *arg, int *continue_walk);
   extern PT_NODE *pt_find_op_type_pre (PARSER_CONTEXT * parser, PT_NODE * node, void *arg, int *continue_walk);
-  extern int pt_get_query_limit_value (PARSER_CONTEXT * parser, PT_NODE * query, DB_VALUE * limit_val);
+  extern int pt_get_query_limit_value (PARSER_CONTEXT * parser, PT_NODE * query, DB_VALUE * limit_val, bool add_offset);
+  extern DB_BIGINT pt_get_query_limit_for_query (PARSER_CONTEXT * parser, PT_NODE * query);
```
Comment (shparkcubrid) [inline src/parser/parser.h]:
```suggestion
  extern DB_BIGINT pt_get_query_limit_from_query (PARSER_CONTEXT * parser, PT_NODE * query);
```

File: src/parser/parser_support.c (hunk 4 above)
Comment (shparkcubrid) [reply src/parser/parser_support.c]: I fixed it.

File: src/parser/parser_support.c (hunk 5 above)
Comment (shparkcubrid) [reply src/parser/parser_support.c:9507]: This function is legacy. I'll keep this function name.

**Review by beyondykk9 — APPROVED** [review]

**Pull request merged**